doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -8,16 +8,176 @@ NEP 35 — Array Creation Dispatching With __array_function__
+different sorts of objects, like dataframes, bags and arrays, here we will focus
+strictly on arrays, which are the objects we can use ``__array_function__``
+with.
+
+Dask uses a graph computing model, meaning it breaks down a large problem in
+many smaller problems and merge their results to reach the final result. To
+break the problem down into smaller ones, Dask also breaks arrays into smaller
+arrays, that it calls "chunks". A Dask array can thus consist of one or more
+chunks and they may be of different types. However, in the context of
+``__array_function__``, Dask only allows chunks of the same type, for example,
+a Dask array can be formed of several NumPy arrays or several CuPy arrays, but
+not a mix of both.
+
+To avoid mismatched types during compute, Dask keeps an attribute ``_meta`` as
+part of its array throughout computation, this attribute is used to both predict
+the output type at graph creation time and to create any intermediary arrays
+that are necessary within some function's computation. Going back to our
+previous example, we can use ``_meta`` information to identify what kind of
+array we would use for padding, as seen below:
+
+.. code:: python
+
+    import numpy as np
+    import cupy
+    import dask.array as da
+    from dask.array.utils import meta_from_array
+
+    def my_pad(arr, padding):
+        padding = np.array(padding, like=meta_from_array(arr))
+        return np.concatenate((padding, arr, padding))
+
+    # Returns dask.array<concatenate, shape=(9,), dtype=int64, chunksize=(5,), chunktype=numpy.ndarray>
+    my_pad(da.arange(5), [-1, -1])
+
+    # Returns dask.array<concatenate, shape=(9,), dtype=int64, chunksize=(5,), chunktype=cupy.ndarray>
+    my_pad(da.from_array(cupy.arange(5)), [-1, -1])
+
+Note how ``chunktype`` in the return value above changes from
+``numpy.ndarray`` in the first ``my_pad`` call to ``cupy.ndarray`` in the
+second.
+
+To enable proper identification of the array type we use Dask's utility function
+``meta_from_array``, which was introduced as part of the work to support
+``__array_function__``, allowing Dask to handle ``_meta`` appropriately. That
+function is primarily targeted at the library's internal usage to ensure chunks
+are created with correct types. Without the ``like=`` argument, it would be
+impossible to ensure ``my_pad`` creates a padding array with a type matching
+that of the input array, which would cause cause a ``TypeError`` exception to
+be raised by CuPy, as discussed above would happen to the CuPy case alone.
+
+Current NumPy users who don't use other arrays from downstream libraries should
+have no impact in their current usage of the NumPy API. In the event of the
+user passing a NumPy array to ``like=``, that will continue to work as if no
+array was passed via that argument. However, this is advised against, as
+internally there will be additional checks required that will have an impact in
+performance.
+
```

Comment:
Move this to the beginning of the section.
```suggestion
NumPy users who don't use other arrays from downstream libraries can
continue to use array creation routines without a `like` argument. Using
``like=ndarray``, will work as if no array was passed via that argument.
However, this will incur additional checks that will negatively impact
performance.

```

Reply:
Done in https://github.com/numpy/numpy/pull/17093/commits/a82cc4b5cb4c3926d2b555d9e8da1146c472a17c .